### ISSUE TITLE
[API Proposal]: System.Text.Json serialization callback context

### ISSUE THREAD

**Issue by IdkGoodName** [dotnet/runtime#59892]
### Background and motivation

Newtonsoft.Json allows you to pass context to its JSON serializer via serializer settings that can later be retrieved in the callbacks. This works well enough for objects that need a specific property (like parent clients, status, etc.) to be set as passed value and it works at any depth. Unfortunately, there isn't a way to do that with System.Text.Json apart from using custom converters, that can't have serializer settings to be passed down (and you can't pass the property value to child types that require same property set if settings aren't passed). This can make it harder to move certain projects over to System.Text.Json and if not make it worse in some instances.

### API Proposal

```C#
namespace System.Text.Json.Serialization
{
    public interface IJsonOnDeserialized<T>
    {
        void OnDeserialized(SomeContextType<T> context);
    }
}
```
or with the breaking change:
```C#
namespace System.Text.Json.Serialization
{
    public interface IJsonOnDeserialized
    {
        void OnDeserialized(StreamingContext context);
    }
}
```

### API Usage

```C#
JsonSerializerOptions options = new JsonSerializerOptions
{
    // ...
    Context = new SomeContextType<int>(10),
    // ... 
};
```
```C#
public class SomeObject : IJsonOnDeserialized<int>
{
    public int Status { get; set; }
 
    public void IJsonOnDeserialized.OnDeserialized(SomeContextType<int> context) => Status = context.Value;
    
    // ... Now we can set up methods or anything else with the received context ...
}
```

### Risks

This could lead to breaking changes if callback interfaces wouldn't use generic types.

**Comment by eiriktsarpalis** [comment]
This was discussed in the serialization callback API proposal, see https://github.com/dotnet/runtime/issues/54528#issuecomment-867802325. I understand passing contexts was cut since it would be difficult to support in the current iteration of source generation.

cc @steveharter

**Comment by eiriktsarpalis** [comment]
This would likely need to be considered in combination with https://github.com/dotnet/runtime/issues/36785.

**Comment by eiriktsarpalis** [comment]
`JsonSerializer` doesn't support passing user state for the moment. This is something we might consider in #71718 and will be implemented in tandem with #63795. Closing as duplicate of #71718.
